### PR TITLE
T18983

### PR DESCRIPTION
**Pull request by smspillaz**
https://phabricator.endlessm.com/T18983

### CODE REVIEW

File: daemon/emer-machine-id-provider.c [hunk 1]
```diff
@@ -291,24 +296,32 @@ emer_machine_id_provider_get_id (EmerMachineIdProvider *self,
...
     emer_machine_id_provider_get_instance_private (self);
 
-  static gboolean id_is_valid = FALSE;
-  G_LOCK_DEFINE_STATIC (id_is_valid);
-
-  G_LOCK (id_is_valid);
-  if (!id_is_valid)
+  /* The assumption is that this will not be called
+   * multiple times for the same instance across different
+   * threads. */
```
Comment (wjt) [inline daemon/emer-machine-id-provider.c]: Is this assumption true? Was the lock not there for a reason before?

I believe the answers are “yes; no”; and so I don't think this comment is necessary. There is plenty of non-thread-safe code in this application.

File: data/eos-metrics-event-recorder.service.in [hunk 2]
```diff
@@ -9,6 +9,7 @@ User=metrics
 Type=dbus
 BusName=com.endlessm.Metrics
 ExecStart=@libexecdir@/eos-metrics-event-recorder
+ExecReload=/bin/kill -SIGHUP $MAINPID
```
Comment (wjt) [inline data/eos-metrics-event-recorder.service.in]: `man systemd.service` actually recommends against this approach – it suggests using a mechanism which allows you to wait for it to complete (and see if it was successful). An example from `dbus.service` itself:

```
/lib/systemd/system/dbus.service:ExecReload=/usr/bin/dbus-send --print-reply --system --type=method_call --dest=org.freedesktop.DBus / org.freedesktop.DBus.ReloadConfig
```

(I suggested the `SIGHUP` approach for `gdm` because it already implemented a `SIGHUP` handler.)

But I appreciate that this approach does not involve adding methods to the API (defined in a separate package) etc etc.

File: daemon/emer-machine-id-provider.c [hunk 3]
```diff
@@ -291,24 +296,32 @@ emer_machine_id_provider_get_id (EmerMachineIdProvider *self,
...
   return TRUE;
 }
+
+void
+emer_machine_id_provider_reload (EmerMachineIdProvider *self)
+{
+  EmerMachineIdProviderPrivate *priv =
+    emer_machine_id_provider_get_instance_private (self);
+
+  g_message ("EmerMachineIdProvider: Reloading from %s", priv->path);
```
Comment (wjt) [inline daemon/emer-machine-id-provider.c]: This message isn't true – that file will not be read until an unspecified time in the future.

File: daemon/emer-daemon.c [hunk 4]
```diff
@@ -1547,6 +1547,15 @@ emer_daemon_new (const gchar             *persistent_cache_directory,
                        NULL);
 }
 
+void
+emer_daemon_reload (EmerDaemon *self)
+{
+  EmerDaemonPrivate *priv = emer_daemon_get_instance_private (self);
+
+  if (priv->machine_id_provider)
```
Comment (wjt) [inline daemon/emer-daemon.c]: I prefer `if (priv->machine_id_provider != NULL)` for checks like this, and a quick skim of this file suggests that's the style in use here.

File: daemon/emer-machine-id-provider.c [hunk 5]
```diff
@@ -165,10 +166,14 @@ emer_machine_id_provider_class_init (EmerMachineIdProviderClass *klass)
                                      emer_machine_id_provider_props);
 }
 
-// Mandatory empty function.
+/* Explicitly set id_is_valid to false */
```
Comment (wjt) [inline daemon/emer-machine-id-provider.c]: I don't think this comment is needed.

File: daemon/emer-machine-id-provider.c [hunk 6]
```diff
@@ -165,10 +166,14 @@ emer_machine_id_provider_class_init (EmerMachineIdProviderClass *klass)
...
 static void
 emer_machine_id_provider_init (EmerMachineIdProvider *self)
 {
+  EmerMachineIdProviderPrivate *priv =
+    emer_machine_id_provider_get_instance_private (self);
+
+  priv->id_is_valid = FALSE;
```
Comment (wjt) [inline daemon/emer-machine-id-provider.c]: I don't really think this is needed – glib guarantees that the priv structure will be 0ed, and the other two fields don't get explicitly initialised here – but ok.

File: data/eos-metrics-event-recorder.service.in (hunk 2 above)
Comment (smspillaz) [reply data/eos-metrics-event-recorder.service.in]: I suppose we could add some new d-bus API to do this, it wouldn't be too hard.

At that point I might as well add some API to wipe the existing metrics cache too.

File: daemon/emer-machine-id-provider.c (hunk 3 above)
Comment (smspillaz) [reply daemon/emer-machine-id-provider.c]: That's true. I suppose I could rephrase it to "will reload from %s"

File: daemon/emer-machine-id-provider.c (hunk 1 above)
Comment (smspillaz) [reply daemon/emer-machine-id-provider.c]: Yeah, the lock being there did seem strange (what was more strange was the fact that we were using a static var for this). In any case, I'm almost certain that you're right - just put that comment there in case anyone would pick up on it ;-)

File: daemon/emer-machine-id-provider.c (hunk 6 above)
Comment (smspillaz) [reply daemon/emer-machine-id-provider.c]: Strictly speaking, that's true, though my preference is to be explicit about these kinds of things.

**Comment by smspillaz** [comment]
@wjt I addressed most of your comments - thanks for the quick feedback :)

File: daemon/emer-main.c [hunk 7]
```diff
@@ -200,6 +209,13 @@ quit_main_loop (GMainLoop *main_loop)
   return G_SOURCE_REMOVE;
 }
 
+static gboolean
+reload_daemon_config (EmerDaemon *daemon)
```
Comment (wjt) [inline daemon/emer-main.c]: I think this function is now unused.

File: daemon/emer-main.c [hunk 8]
```diff
@@ -222,6 +238,8 @@ on_bus_acquired (GDBusConnection *system_bus,
                     G_CALLBACK (on_set_enabled), daemon);
   g_signal_connect (server, "handle-upload-events",
                     G_CALLBACK (on_upload_events), daemon);
+  g_signal_connect (server, "handle-reload-config",
+                    G_CALLBACK (on_reload_config), daemon);
```
Comment (wjt) [inline daemon/emer-main.c]: Should this be a privileged operation, ie. added to `on_authorize_method_check`? You could make an argument that calling this method has no observable effect if `/etc/machine-id` hasn't changed, so it does not need to be authorized.

**Comment by smspillaz** [comment]
Okay, reworked the approach here considering that we can't just nuke the machine-id.

I added a `ResetTrackingId` method to the daemon which calls through a system helper to write an override file in `/etc/eos-event-recorder-daemon/machine-id-override`. If that file is present, its value is preferred to that of `/etc/machine-id`. This will allow the demo mode to reset the machine-id safely.

File: system-helper/eos-event-recorder-system-helper.c [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+#include <glib-unix.h>
+#include <polkit/polkit.h>
+
+#include <uuid/uuid.h>
+
+#include <shared/metrics-util.h>
+
+#include "emer-metrics-system-helper.h"
+
+#define TRACKING_ID_OVERRIDE SYSCONFDIR "/eos-metrics-event-recorder/machine-id-override"
```
Comment (wjt) [inline system-helper/eos-event-recorder-system-helper.c]: A better choice of path would be `/etc/metrics/machine-id-override`. The daemon already writes a file to this directory ([`cache-size.conf`](https://github.com/endlessm/eos-event-recorder-daemon/blob/master/daemon/emer-cache-size-provider.c#L39); `eos-metrics-permissions.conf` is also in that directory but for reasons that elude me it is edited by external scripts and re-read by the daemon every time its contents are needed). The `metrics` user owns this directory so there is no need for a privileged helper.

Comment (smspillaz) [reply system-helper/eos-event-recorder-system-helper.c]: Ah of course! That would allow me to drop the system helper too...

File: Makefile.am [hunk 10]
```diff
@@ -55,6 +55,33 @@ noinst_PROGRAMS =
...
+# Binaries that should be installed in libexecdir
+libexec_PROGRAMS = 
+
+substitute_libexecdir = sed \
+	-e 's|@libexecdir[@]|$(libexecdir)|g' \
+	$(NULL)
+
+# systemd service units
+#
+# Defined as systemdsystemunitdir in ../configure.ac.
```
Comment (wjt) [inline Makefile.am]: The relative path in this comment is now wrong since this file is in the same directory as `configure.ac`.

File: daemon/emer-machine-id-provider.c [hunk 11]
```diff
@@ -78,23 +96,23 @@ static GParamSpec *emer_machine_id_provider_props[NPROPS] = { NULL, };
...
 
-  return priv->path;
+  return priv->paths;
 }
 
 static void
-set_id_path (EmerMachineIdProvider *self,
-             const gchar           *given_path)
+set_id_paths (EmerMachineIdProvider *self,
+              const GStrv            given_paths)
```
Comment (wjt) [inline daemon/emer-machine-id-provider.c]: If you make this take ownership a non-const `GStrv`…

File: daemon/emer-machine-id-provider.c [hunk 12]
```diff
@@ -224,62 +250,118 @@ hyphenate_uuid (gchar *uuid_sans_hyphens)
...
+    }
+
+  if (uuid_is_null (id))
+    {
+      g_critical ("Failed to read in a unique machine id");
+      return FALSE;
+    }
+
+  uuid_copy (priv->id, id);
+  uuid_clear (id);
```
Comment (wjt) [inline daemon/emer-machine-id-provider.c]: This `uuid_clear` is not needed, it's just a stack-allocated array.

File: tests/daemon/test-machine-id-provider.c [hunk 13]
```diff
@@ -31,7 +31,9 @@
 #define HYPHENS_IN_ID 4
 
 #define TESTING_FILE_PATH "/tmp/testing-machine-id"
+#define TESTING_OVERRIDE_FILE_PATH "/tmp/override-testing-machine-id"
```
Comment (wjt) [inline tests/daemon/test-machine-id-provider.c]: I appreciate the existing test does the same but I don't really like using fixed paths in `/tmp` here. I think the fixture setup should create the temporary files.

File: tests/daemon/test-machine-id-provider.c [hunk 14]
```diff
@@ -31,7 +31,9 @@
...
 #define TESTING_ID        "04448f74fde24bd7a16f8da17869d5c3\n"
+#define TESTING_OVERRIDE_ID        "04448f74fde24bd7a16f8da17869d5c4\n"
```
Comment (wjt) [inline tests/daemon/test-machine-id-provider.c]: Might be nice to generate a more-distinct UUID so that assertion failures are clearer, rather than one which differs in only 2 trailing bits!

File: tests/daemon/test-machine-id-provider.c [hunk 15]
```diff
@@ -102,8 +106,55 @@ static void
...
+    NULL
+  };
   EmerMachineIdProvider *id_provider =
-    emer_machine_id_provider_new_full (TESTING_FILE_PATH);
+    emer_machine_id_provider_new_full (paths);
+  uuid_t id;
+  g_assert (emer_machine_id_provider_get_id (id_provider, id));
+  gchar unparsed_id[HYPHENS_IN_ID + FILE_LENGTH];
+  uuid_unparse_lower (id, unparsed_id);
+  gchar* unhypenated_id = unhyphenate_uuid (unparsed_id);
```
Comment (wjt) [inline tests/daemon/test-machine-id-provider.c]: g_autofree here and remove g_free below

File: tests/daemon/test-machine-id-provider.c [hunk 16]
```diff
@@ -102,8 +106,55 @@ static void
...
+  g_assert_cmpstr (TESTING_ID, ==, unhypenated_id);
+  g_free (unhypenated_id);
+  g_object_unref (id_provider);
```
Comment (wjt) [inline tests/daemon/test-machine-id-provider.c]: If you were feeling keen you could add `G_DEFINE_AUTOPTR_CLEANUP_FUNC (EmerMachineIdProvider, g_object_unref);` to `emer-machine-id-provider.h` and then use `g_autoptr(EmerMachineIdProvide) id_provider` above.

File: tests/daemon/test-machine-id-provider.c [hunk 17]
```diff
@@ -128,6 +179,10 @@ main (gint                argc,
                        test_machine_id_provider_new_succeeds);
   ADD_CACHE_TEST_FUNC ("/machine-id-provider/can-get-id",
                        test_machine_id_provider_can_get_id);
+  ADD_CACHE_TEST_FUNC ("/machine-id-provider/can-get-id-override",
+                       test_machine_id_provider_can_get_id_override);
+  ADD_CACHE_TEST_FUNC ("/machine-id-provider/can-get-id-override-not-exist",
+                       test_machine_id_provider_can_get_id_override_not_exist);
```
Comment (wjt) [inline tests/daemon/test-machine-id-provider.c]: Please also test the case where the first file exists but is malformed.

File: Makefile.am [hunk 18]
```diff
@@ -62,11 +62,6 @@ substitute_libexecdir = sed \
 	-e 's|@libexecdir[@]|$(libexecdir)|g' \
 	$(NULL)
 
-# systemd service units
-#
-# Defined as systemdsystemunitdir in ../configure.ac.
-dist_systemdsystemunit_DATA =
```
Comment (wjt) [inline Makefile.am]: This  was only just moved to `Makefile.am` two patches ago!

File: Makefile.am [hunk 19]
```diff
@@ -77,7 +72,7 @@ dist_systembussecuritypolicy_DATA =
 
 # policykit
 policydir = ${datadir}/polkit-1/actions
-dist_policy_DATA = data/com.endlessm.Metrics.policy
```
Comment (wjt) [inline Makefile.am]: Should this not be part of the previous patch?

File: Makefile.am [hunk 20]
```diff
@@ -86,6 +81,9 @@ persistentcachedir=$(localstatedir)/cache/metrics/
 configdir=$(sysconfdir)/metrics/
 permissions_file=$(configdir)eos-metrics-permissions.conf
 
+# systemd
+systemdsystemunit_DATA =
```
Comment (wjt) [inline Makefile.am]: What's this doing here?

File: tests/daemon/test-daemon.c [hunk 21]
```diff
@@ -77,6 +78,8 @@ typedef struct _Fixture
   gint64 relative_time;
   gint64 absolute_time;
   gchar *request_path;
+
+  gchar *tempfiles_dir;
```
Comment (wjt) [inline tests/daemon/test-daemon.c]: This appears to be unused.

File: tests/daemon/test-daemon.c [hunk 22]
```diff
@@ -924,13 +927,17 @@ create_test_object (Fixture *fixture)
...
+                          100000 /* max bytes buffered */,
+                          NULL /* override machine id path */);
 }
 
 static void
 setup (Fixture      *fixture,
        gconstpointer unused)
 {
+  g_autoptr(GFileIOStream) stream = NULL;
+  g_autoptr(GError) error = NULL;
```
Comment (wjt) [inline tests/daemon/test-daemon.c]: These variables are not used.

File: daemon/emer-daemon.c [hunk 23]
```diff
@@ -1547,6 +1580,29 @@ emer_daemon_new (const gchar             *persistent_cache_directory,
                        NULL);
 }
 
+gboolean
+emer_daemon_reset_tracking_id (EmerDaemon  *self,
+                               GError     **error)
+{
+  EmerDaemonPrivate *priv = emer_daemon_get_instance_private (self);
+
+  if (priv->override_machine_id_path != NULL)
```
Comment (wjt) [inline daemon/emer-daemon.c]: Given the point of `EmerMachineIdProvider` is to abstract away how the machine ID is stored, I think it should have an `emer_machine_id_provider_reset_tracking_id` function. It already knows what the override path is, and the call to `emer_machine_id_provider_reload` to tell it the world has changed underneath its feet suggests that it should take care of the reset, too. This would also make the test case simpler because you don't need to hit the disk at all, and you can just assert that the mock version of this function was called.

File: daemon/emer-daemon.c [hunk 24]
```diff
@@ -1547,6 +1580,29 @@ emer_daemon_new (const gchar             *persistent_cache_directory,
...
+  else
+    {
+      if (!write_tracking_id_file (TRACKING_ID_OVERRIDE, error))
+        return FALSE;
+    }
+
+  if (priv->machine_id_provider != NULL)
+    emer_machine_id_provider_reload (priv->machine_id_provider);
+
+  return TRUE;
```
Comment (wjt) [inline daemon/emer-daemon.c]: If there were a signal that `eos-metrics-instrumentation` could use to discard all its state and re-emit the boot-time events, then I think the `ResetTrackingId` method should discard all buffered events and then emit that signal. But it's both or neither: if we discard all cached events here, we will receive the "it's demo mode!" event (which is emitted from g-i-s only after this method returns) but not the boot-time events with info like the OS version, image, hardware, etc.

Comment (wjt) [reply daemon/emer-daemon.c]: (I'm not proposing adding that signal and behaviour here – I think that's a separate task.)

File: daemon/emer-machine-id-provider.c [hunk 25]
```diff
@@ -60,12 +64,26 @@ G_DEFINE_TYPE_WITH_PRIVATE (EmerMachineIdProvider, emer_machine_id_provider, G_T
...
   PROP_0,
-  PROP_PATH,
+  PROP_PATHS,
   NPROPS
 };
 
 static GParamSpec *emer_machine_id_provider_props[NPROPS] = { NULL, };
 
+static GStrv
+dup_strv (GStrv input)
```
[the remark this answers is not in the document]
Comment (smspillaz) [reply daemon/emer-machine-id-provider.c]: Ah damn, I was looking for that and must have missed it. Thanks!

File: Makefile.am (hunk 18 above)
Comment (smspillaz) [reply Makefile.am]: `git rebase` 😿😿😿😿

File: daemon/emer-daemon.c (hunk 23 above)
Comment (smspillaz) [reply daemon/emer-daemon.c]: This is certainly a tension I was feeling too when implementing this. `EmerMachineIdProvider` takes a strv of potential override paths in priority order, so it can't really know which path to write to.

That is unless what we want to do is only have a single override path, which could be a smart move.

File: Makefile.am (hunk 20 above)
Comment (smspillaz) [reply Makefile.am]: We append to it later in one of the subdirectories included by this makefile

File: tests/daemon/test-machine-id-provider.c [hunk 26]
```diff
@@ -72,25 +80,41 @@ unhyphenate_uuid (gchar *uuid_with_hyphens)
...
+  g_free (fixture->test_temp_path);
 }
 
 // Testing Cases
 
 static void
-test_machine_id_provider_new_succeeds (gboolean     *unused,
-                                       gconstpointer dontuseme)
+test_machine_id_provider_new_succeeds (MachineIdTestFixture *fixture,
+                                       gconstpointer          dontuseme)
```
Comment (wjt) [inline tests/daemon/test-machine-id-provider.c]: Mismatched alignment

File: daemon/emer-machine-id-provider.c [hunk 27]
```diff
@@ -156,24 +169,35 @@ emer_machine_id_provider_class_init (EmerMachineIdProviderClass *klass)
 
   /* Blurb string is good enough default documentation for this */
   emer_machine_id_provider_props[PROP_PATH] =
-    g_param_spec_string ("path", "Path",
-                         "The path to the file where the unique identifier is stored.",
+    g_param_spec_string ("path", "Machine ID path",
+                         "The path to where the immutable machine id is stored.",
```
Comment (wjt) [inline daemon/emer-machine-id-provider.c]: `id` here should be capitalised

File: daemon/emer-machine-id-provider.c [hunk 28]
```diff
@@ -156,24 +169,35 @@ emer_machine_id_provider_class_init (EmerMachineIdProviderClass *klass)
...
                          DEFAULT_MACHINE_ID_FILEPATH,
                          G_PARAM_WRITABLE | G_PARAM_CONSTRUCT_ONLY | G_PARAM_STATIC_STRINGS);
+  emer_machine_id_provider_props[PROP_OVERRIDE_PATH] =
+    g_param_spec_string ("override-path", "Tracking ID override path",
+                         "The path to where a mutable tracking id override path is stored.",
```
Comment (wjt) [inline daemon/emer-machine-id-provider.c]: Here too

File: daemon/emer-machine-id-provider.c [hunk 29]
```diff
@@ -224,62 +254,119 @@ hyphenate_uuid (gchar *uuid_sans_hyphens)
...
+                      priv->override_path,
+                      local_error->message);
+        }
+
+      g_clear_error (&local_error);
+
+      if (!read_one_machine_id (priv->path, id, &local_error))
+        {
+          g_message ("Failed to read machine id %s: %s",
+                      priv->override_path,
```
Comment (wjt) [inline daemon/emer-machine-id-provider.c]: This is the wrong path.

File: daemon/emer-machine-id-provider.c [hunk 30]
```diff
@@ -224,62 +250,118 @@ hyphenate_uuid (gchar *uuid_sans_hyphens)
...
+            }
+          continue;
+        }
+
+      break;
+    }
+
+  if (uuid_is_null (id))
+    {
+      g_critical ("Failed to read in a unique machine id");
```
[the remark this answers is not in the document]
Comment (wjt) [reply daemon/emer-machine-id-provider.c]: I still think a better message would be something like:

> Failed to read machine ID from either %s or %s

for the two paths (override, then fallback)

File: daemon/emer-machine-id-provider.c (hunk 12 above)
Comment (wjt) [reply daemon/emer-machine-id-provider.c]: Still true!

File: daemon/emer-machine-id-provider.c [hunk 31]
```diff
@@ -291,24 +378,64 @@ emer_machine_id_provider_get_id (EmerMachineIdProvider *self,
...
+
+  if (!g_file_make_directory_with_parents (directory, NULL, &local_error))
+    {
+      if (!g_error_matches (local_error, G_IO_ERROR, G_IO_ERROR_EXISTS))
+        {
+          g_propagate_error (error, g_steal_pointer (&local_error));
+          return FALSE;
+        }
+
+      g_clear_error (&local_error);
```
Comment (wjt) [inline daemon/emer-machine-id-provider.c]: Unneeded since `local_error` is not re-used

File: daemon/emer-machine-id-provider.c [hunk 32]
```diff
@@ -291,24 +378,64 @@ emer_machine_id_provider_get_id (EmerMachineIdProvider *self,
...
+{
+  EmerMachineIdProviderPrivate *priv =
+    emer_machine_id_provider_get_instance_private (self);
+
+  if (!write_tracking_id_file (priv->override_path, error))
+    return FALSE;
+
+  g_message ("EmerMachineIdProvider: Will reload from:\n%s", priv->override_path);
+
+  priv->id_is_valid = FALSE;
```
Comment (wjt) [inline daemon/emer-machine-id-provider.c]: This function needs to `return TRUE` on the happy path. I'd hope the compiler would issue a warning.

File: daemon/emer-machine-id-provider.c [hunk 33]
```diff
@@ -291,24 +378,64 @@ emer_machine_id_provider_get_id (EmerMachineIdProvider *self,
...
+emer_machine_id_provider_reset_tracking_id (EmerMachineIdProvider  *self,
+                                            GError                **error)
+{
+  EmerMachineIdProviderPrivate *priv =
+    emer_machine_id_provider_get_instance_private (self);
+
+  if (!write_tracking_id_file (priv->override_path, error))
+    return FALSE;
+
+  g_message ("EmerMachineIdProvider: Will reload from:\n%s", priv->override_path);
```
Comment (wjt) [inline daemon/emer-machine-id-provider.c]: I think the newline should be removed

File: data/com.endlessm.Metrics.policy [hunk 34]
```diff
@@ -38,4 +38,14 @@
     </defaults>
   </action>
 
+  <action id="com.endlessm.Metrics.ResetTrackingId">
```
Comment (wjt) [inline data/com.endlessm.Metrics.policy]: Is this used? I don't see any reference to it. I think you need to update `on_authorize_method_check`.

File: data/com.endlessm.Metrics.policy [hunk 35]
```diff
@@ -38,4 +38,14 @@
...
+    <description>Reset the tracking id used by the metrics system</description>
```
Comment (wjt) [inline data/com.endlessm.Metrics.policy]: ID should be capitalized

**Comment by wjt** [comment]
You'll also need to rebase the branch because it now conflicts with master.

File: daemon/emer-machine-id-provider.c (hunk 32 above)
Comment (smspillaz) [reply daemon/emer-machine-id-provider.c]: It does not, unfortunately, I have noticed :(

File: data/com.endlessm.Metrics.policy (hunk 34 above)
Comment (smspillaz) [reply data/com.endlessm.Metrics.policy]: Ah! looks like I overzealously reverted this change. I'll unrevert it, since we should make resetting the tracking ID a privileged operation.

**Comment by smspillaz** [comment]
Addressed all comments. Also looks like a file was missing on master, so I added that.

Should be able to squash down all the commits into 9eeba0d once we're done, just included them here for completeness.

File: daemon/emer-main.c [hunk 36]
```diff
@@ -134,6 +151,58 @@ on_upload_events (EmerEventRecorderServer *server,
...
+
+static const gchar *not_authorized_error_messages[] = {
+  NULL,
+  "Disabling metrics is only allowed from system settings",
+  "Only privileged users can reset the tracking id",
+};
+static const gsize n_not_authorized_error_messages = G_N_ELEMENTS (not_authorized_error_messages);
+
+static const gchar *
+not_authorized_error_message_for_method (AuthorizedMethod method)
```
Comment (wjt) [inline daemon/emer-main.c]: This method is never used!

I think a better approach here would be to define:

```c
typedef struct {
    const gchar *method_name;
    const gchar *method_full_name;
    const gchar *error_message;
} AuthorizedMethod;

static const AuthorizedMethods[] authorized_methods = {
   { "SetEnabled", "com.endlessm.Metrics.SetEnabled", "..." },
   { "ResetTrackingId", ... },
   { NULL }
};

static const AuthorizedMethod *
lookup_authorized_method (const gchar *method_name)
{
  // obvious loop, return NULL if not found
}
```

Then `on_authorize_method_check` would call this function once, early-return if it returned NULL, and then have all the other information it needed right there. No need for extra lookups and `g_assert`s.
```

Comment (smspillaz) [reply daemon/emer-main.c]: Makes sense, I think that's a lot simpler.

**Comment by smspillaz** [comment]
Okay, rebased on #225, squashed all the commits except the last one with the last review comment together.

File: daemon/emer-main.c [hunk 37]
```diff
@@ -186,8 +239,8 @@ on_authorize_method_check (GDBusInterfaceSkeleton *interface,
     g_dbus_method_invocation_return_error (invocation,
                                            G_DBUS_ERROR,
                                            G_DBUS_ERROR_AUTH_FAILED,
-                                           "Disabling metrics is only "
-                                           "allowed from system settings");
+                                           "%s",
+                                           authorized_method->error_message);
```
Comment (wjt) [inline daemon/emer-main.c]: If you were feeling really keen you could use `g_dbus_method_invocation_return_error_literal` and avoid the `"%s"`.

File: Makefile.am [hunk 38]
```diff
@@ -98,6 +126,9 @@ EOS_COVERAGE_BLACKLIST_PATTERNS =	\
 AM_CFLAGS += @EOS_C_COVERAGE_CFLAGS@
 AM_LDFLAGS += @EOS_C_COVERAGE_LDFLAGS@
 
+# # # COMMON LIBRARY # # #
+include $(top_srcdir)/shared/Makefile.am.inc
```
Comment (wjt) [inline Makefile.am]:
```
automake: error: cannot open < shared/Makefile.am.inc: No such file or directory
autoreconf: automake failed with exit status: 1
```

Comment (smspillaz) [reply Makefile.am]: Ouch, thanks for spotting - I rebased and put it into the right commit.

File: tests/daemon/mock-machine-id-provider.c [hunk 39]
```diff
@@ -48,16 +55,29 @@ emer_machine_id_provider_new (void)
...
   return TRUE;
 }
+
+gboolean
+emer_machine_id_provider_reset_tracking_id (EmerMachineIdProvider  *self,
+                                            GError                **error)
+{
+  EmerMachineIdProviderPrivate *priv =
+    emer_machine_id_provider_get_instance_private (self);
+  priv->has_override = TRUE;
```
Comment (wjt) [inline tests/daemon/mock-machine-id-provider.c]:
```
../../tests/daemon/mock-machine-id-provider.c: In function 'emer_machine_id_provider_reset_tracking_id':
../../tests/daemon/mock-machine-id-provider.c:83:1: error: control reaches end of non-void function [-Werror=return-type]
```

Will fix this…
